### DISCUSSION TITLE
zodios cause big bundle size (on nextjs)

### DISCUSSION THREAD

**Discussion by QzCurious** [ecyrbe/zodios#197] [Q&A] [answered]
I'm not familiar how bundler works and how to trace it down. I am willing to inspect it more, please give me some instructions.

## The problem

I'm using nextjs. When integrate with zodios, a simple request can cause build bundle up to 197KB. Although `@zodios/core` is only of size 3.7 kb. I don't know if it's me importing thing wrong. Please explain to me what is causing this to be happend.
![image](https://user-images.githubusercontent.com/38932402/197252507-29ea36cc-8b93-4a2d-b81b-5411657294fe.png)

Here is a [repo](https://github.com/QzCurious/zodios-cause-big-bundle-size) to showcase. Each commit take a small step for reviewing how code and bundle size changes. I've also configured it with `@next/bundle-analyzer` so you can `ANALYZE=true yarn build` to inspect the bundle (but I can't tell how and what is wrong, sadly).

In summary:

| clean page | with axios | with zod | with zodios |
| :--------: | :--------: | :------: | :---------: |
|  77.9 kB   |  95.3 kB   |  106 kB  |   197 kb    |

One wierd thing I noticed. bundle-analyzer list out a few packages that is not mean to be ship to production are included in bundle. 
![image](https://user-images.githubusercontent.com/38932402/197252195-6df33306-348c-4b4f-9cb8-2f0ccf558e9d.png)

**Comment by ecyrbe** [comment]
Can you analyse it further. Can you show me the output of bundle analyser for each steps?

**Reply by QzCurious** [reply]
I've break down each step to a commit. We can checkout each commit and run `yarn build`. nextjs will print out the bundle sizes.

Here is how to inspect the repo:
![image](https://user-images.githubusercontent.com/38932402/197316664-c11b77a1-7bef-4a95-955c-07eec6ee115b.png)

On left hand side, a commit (e8847eb) with zodios on index page is size 197 kb.
On right hand side, a commit (d719ffa) with all zodios dependencies (by without zodios) on index page is size 106 kB.

If you want to have a more precise view. Do `ANALYZE=true yarn build` instead.

**Reply by QzCurious** [reply]
I'm not sure how to analyse it more. Seem like you have some clue about the cause. It would be nice if you could point me out how you inspecting it. So, if I can, I can provide more helpful information on futrue. Thank you.

**Reply by QzCurious** [reply]
One thing I confused more on output of `@next/bundle-analyzer`. Why the extra pieces of code is not under zodios but next/dist. It's very hard for me to track down the problem is more likely caused by zodios.

**Comment by ecyrbe** [comment]
@QzCurious i found the issue.
Internally zodios has a formdata plugin.
- On Browser: zodios is using Native Browser FormData
- On NodeJs: zodios is using a node form-data lib

And it seems like nextjs webpack bundler does not understand zodios declaration to not use node form-data lib on Browser build.
It then add all polyfills to make the form-data lib working on the browser.

It's plain bad. i'll try to figure out a solution for this.

So i'm converting this discussion to an issue

**Reply by ecyrbe** [reply]
#198

**Comment by ecyrbe** [comment, marked as answer]
@QzCurious A new version of @zodios/core and @zodios/react to 10.3.0 just landed and fixes this issue with esm build now also supported.

**Reply by QzCurious** [reply]
Confirmed that. And the [repo](https://github.com/QzCurious/zodios-cause-big-bundle-size) is updated with steps to confirm it.

In summary:

| clean page | with axios | with zod | with @zodios/core@10.2.0 | with @zodios/core@10.3.0 | with @zodios/react@10.3.1 |
| :--------: | :--------: | :------: | :----------------------: | :----------------------: | :-----------------------: |
|  77.9 kB   |  95.3 kB   |  106 kB  |          197 kb          |          108 kb          |          120 kb           |

(note that **with @zodios/react@10.3.1** includes its depencdency @tanstack/react-query so the extra kb)
